File: Java/017_LetterCombinationsOfAPhone.java
public class Solution {
    String[] alpha = new String[] {
            " ",
            "1", "abc", "def",
            "ghi", "jkl", "mno",
            "pqrs", "tuv", "wxyz"
    };
    
    StringBuilder word;

    private void dfs(List<String> res, String digits, int cur) {
        if (cur >= digits.length()) {
            res.add(word.toString());
        } else {
            for (int i = 0; i < alpha[digits.charAt(cur) - '0'].length(); ++i) {
                word.append(alpha[digits.charAt(cur) - '0'].charAt(i));
                dfs(res, digits, cur + 1);
                //将最后一位清楚，进行下一位处理
                word.deleteCharAt(word.length() - 1);
            }
        }
    }

    public List<String> letterCombinations(String digits) {
        List<String> ret = new ArrayList<String>();
        
        if (digits.length() == 0) {
            return ret;
        }
        
        word = new StringBuilder();
        dfs(ret, digits, 0);
        return ret;
    }
}

File: Python/017_LetterCombinationsOfAPhone.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        alpha = [" ", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"] 
        res = []
        word = []
        
        def dfs(cur):
            if cur >= len(digits):
                res.append(''.join(word))
            else:
                for x in alpha[(int)(digits[cur]) - (int)('0')]:
                    word.append(x)
                    dfs(cur + 1)
                    word.pop()
                    
        if len(digits) == 0:        
            return res
        else:
            dfs(0)
            return res